### ISSUE TITLE
Support non-automatic loading of keys for encrypted datasets on boot

### ISSUE THREAD

**Issue by thedeadliestcatch** [openzfs/zfs#15840]
### Describe the feature would like to see added to OpenZFS

At the moment, all encrypted datasets are mounted/have their keys loaded automatically on boot time. This interrupts boot when the key is not available or a passphrase is required to "open" the dataset.

Ideally, some property should be added, ex. "noauto", so that the key is only manually loaded and the dataset is ignored during boot.

### How will this feature improve OpenZFS?

It will prevent systems with encryption-at-rest from having the boot process interrupted when the keys are loaded via a manual or more complex system that depends on external tooling.

### Additional context

This can be observed by creating a password-based encrypted dataset in a system coexisting with datasets encrypted using key files.

**Comment by robn** [comment]
Is it enough to set `canmount=off` or `canmount=noauto` on the datasets you don't want mounted?

**Comment by robn** [comment]
Right, I see. That's OpenZFS' `etc/init.d/zfs-load-key.in`, which of course distros are under no obligation to use but at least we could make some change if necessary.

It looks like that's doing `load-key` if `keylocation` is _not_ `prompt` or `none`, that is, it asserts a real location. Is that right?

If that's the case, and you do have a manual/external process, I think probably you should set `keylocation=none` and then in your separate process, use `zfs load-key -L <keylocation>` to specify the location directly at the proper time. It seems like it doesn't make a lot of sense to add a special property to say "don't do this thing I asked you to do".

If that's _not_ the case, well, tell me more!

**Comment by voidzero** [comment]
It's not possible to set `keylocation=none` on an encrypted filesystem:

```
# zfs set keylocation=none zroot/enc
cannot set property for 'zroot/enc': keylocation must not be 'none' for encrypted datasets
```

Setting it to `canmount=noauto` would be my preferred way to skip the automatic loading of the filesystem at boot, but on second thought this might make the `zfs mount -a` command ambiguous.

Currently, any linux distribution that loads all the filesystems with `zfs load-key -a` cannot be told to skip any individual filesystem. I'd be very happy if this can be overridden from zfs in some way - I want a prompt for some filesystems, but not load them at boot time.

**Comment by voidzero** [comment]
I just discovered that NixOS has the following setting:
[boot.zfs.requestEncryptionCredentials](https://search.nixos.org/options?channel=24.05&from=0&size=50&sort=relevance&type=packages&query=boot.zfs.requestEncryptionCredentials)

That really made my day.
